src/main/scala/com/microsoft/hyperspace/actions/RefreshIncrementalAction.scala

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.actions
+
+import org.apache.spark.sql.{SaveMode, SparkSession}
+import org.apache.spark.sql.functions.col
+
+import com.microsoft.hyperspace.HyperspaceException
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.DataFrameWriterExtensions.Bucketizer
+import com.microsoft.hyperspace.telemetry.{AppInfo, HyperspaceEvent, RefreshIncrementalActionEvent}
+
+/**
+ * Action to create indexes on newly arrived data. If the user appends new data to existing,
+ * pre-indexed data, they can use refresh api to generate indexes only on the additional data.
+ *
+ * Algorithm Outline:
+ * - Identify newly added data files.
+ * - Create new index version on these files.
+ * - Update metadata to reflect the latest snapshot of index. This snapshot includes all the old
+ *   and the newly created index files. The source content points to the latest data files.
+ *
+ * @param spark SparkSession.
+ * @param logManager Index LogManager for index being refreshed.
+ * @param dataManager Index DataManager for index being refreshed.
+ */
+class RefreshIncrementalAction(
+    spark: SparkSession,
+    logManager: IndexLogManager,
+    dataManager: IndexDataManager)
+    extends RefreshActionBase(spark, logManager, dataManager) {
+  final override def op(): Unit = {
+    logInfo(
+      "Refresh index is updating index by removing index entries " +
+        s"corresponding to ${deletedFiles.length} deleted source data files.")
+
+    if (appendedFiles.nonEmpty) {
```

Comment:
Good point. Previously, with only filenames instead of fileInfos, RefreshDeleteAction is required to be done firstly. 
Because if we do RefreshAppendAction first, we cannot distinguish which rows should be removed while RefreshDeleteAction. 

Now we don't need to care about the order because we know the list of index data files not including appended data.
I changed the order just because AppendRefresh uses `write` function which uses `overwrite` mode to write the data.
(I tried to keep previous impl. mostly)

Either way, mode: SaveMode is required because there are 2 data writings for "delete" and "append" respectively.

Reply:
I added some code to use "overwrite" mode in case there is no appended files, to make sure everything is clear before the write.
Thanks :)